File: C/file_reading.cpp
#include <stdio.h>
#include <iostream>
#include <fstream>
#include <string>
#include <math.h>
#include "global_variables.h"


int n_nodes;
int n_triangles;
int n_edges;
int n_neigh;
int n_traces;
int n_trace_vertex;
double accuracy = pow(10,-14);

std::vector<NODE> node;
std::vector<std::vector<int> > triangle;
std::vector<std::vector<int> > edge;
std::vector<std::vector<int> > neigh;
std::vector<std::vector<double> > trace_vertex;
std::vector<std::vector<int> > trace;
std::vector<std::vector<double> > T;
std::vector<double> T_squared;
std::vector<double> toll_t;
std::vector<std::vector<int> > queue;

void load_max_numbers() {

  //-------PRENDO N_NODES----------
  std::ifstream input;
  input.open("quad.1.node.txt");

  if (!input) {
    std::cout << "Cannot open file.\n";
    return;
  }
  input >> n_nodes;

  input.close();

  //-------PRENDO N_TRIANGLES----------
  std::ifstream input1;
  input1.open("quad.1.ele.txt");

  if (!input1) {
    std::cout << "Cannot open file.\n";
    return;
  }
  input1 >> n_triangles;
  input1.close();

  //-------PRENDO N_EDGES----------
  std::ifstream input2;
  input2.open("quad.1.edge.txt");

  if (!input2) {
    std::cout << "Cannot open file.\n";
    return;
  }
  input2 >> n_edges;
  input2.close();

  //-------PRENDO N_NEIGH----------
  std::ifstream input3;
  input3.open("quad.1.neigh.txt");

  if (!input3) {
    std::cout << "Cannot open file.\n";
    return;
  }
  input3 >> n_neigh;
  input3.close();


    //-------PRENDO N_TRACES e N_TRACES_VERTEX----------
    std::ifstream input4;
    input4.open("quad.1.neigh.txt");

    if (!input4) {
      std::cout << "Cannot open file.\n";
      return;
    }
    input4 >> n_trace_vertex;
    input4 >> n_traces;
    input4.close();

};


void load_matrices(){

  // variabili temporanee per mangiare righe e valori
  std::string sGarbage;
  int garbage;

  //--------RIEMPO NODE----------------
  // matrice di vettori, definita in global_variables.h
  node.resize(n_nodes);
  std::ifstream input;
  input.open("quad.1.node.txt");

  if (!input) {
    std::cout << "Cannot open file.\n";
    return;
  }
  getline(input,sGarbage);
  for(int i = 0; i<n_nodes; i++){
    input >> garbage;
    input >> node[i].x;
    input >> node[i].y;
    input >> garbage;
    node[i].side = 0;
    node[i].toll = -1;
    node[i].edges[0] = 0;
    node[i].edges[1] = 0;
    node[i].edges[2] = 0;
    node[i].triangles[0] = 0;
    node[i].triangles[1] = 0;
    node[i].tot_edges = 0;
    node[i].tot_triangles = 0;
    node[i].s = 666;
  }
  input.close();

  //------RIEMPO TRIANGLE-----------
  // triangle(n_triangles,7)
  // prime tre colonne indici nodi, seconde tre colonne status edge,
  // ultima colonna flag
  triangle.resize(n_triangles); // faccio crescere le righe
  for(int j = 0; j<n_triangles; j++){
    triangle[j].resize(7); // faccio crescere le colonne per ogni riga
  }

  std::ifstream input1;
  input1.open("quad.1.ele.txt");

  if (!input1) {
    std::cout << "Cannot open file.\n";
    return;
  }
  getline(input1,sGarbage); // mangio prima riga del file
  int j = 0;
  for(int i = 0; i<n_triangles; i++){
    input1 >> garbage;
    input1 >> triangle[i][j];
    input1 >> triangle[i][j+1];
    input1 >> triangle[i][j+2];
    triangle[i][j+3] = -1;
    triangle[i][j+4] = -1;
    triangle[i][j+5] = -1;
    triangle[i][j+6] = -1;
    }
  input1.close();

//------RIEMPO EDGE--------
// edge(n_edges,3)
// edge(n,1) vertice 1 del segmento n
// edge(n,2) vertice 2 del segmento n
// edge(n,3) status con la traccia
edge.resize(n_edges); // faccio crescere righe
for (int i = 0; i<n_edges; i++){
  edge[i].resize(3); // per ogni riga faccio crescere la colonna
}

std::ifstream input2;
input2.open("quad.1.edge.txt");

if (!input2) {
  std::cout << "Cannot open file.\n";
  return;
}

getline(input2,sGarbage);
for(int i = 0; i < n_edges; i++){
  input2 >> garbage;
  input2 >> edge[i][0];
  input2 >> edge[i][1];
  input2 >> garbage;
  edge[i][2] = -1;
}
input2.close();

//-------RIEMPO NEIGH--------
// neigh(n_neigh,3)
// neigh(5,2) vicino 2 del triangolo 5, -1 default per non vicino.

neigh.resize(n_neigh); // faccio crescere righe
for(int i = 0; i<n_neigh; i++){
  neigh[i].resize(3); // faccio crescere colonne;
}

std::ifstream input3;
input3.open("quad.1.neigh.txt");

if (!input3) {
  std::cout << "Cannot open file.\n";
  return;
}

getline(input3,sGarbage);
for(int i = 0; i<n_neigh; i++){
    input3 >> garbage;
    input3 >> neigh[i][j];
    input3 >> neigh[i][j+1];
    input3 >> neigh[i][j+2];
}
input3.close();

//------RIEMPO TRACE_VERTEX E TRACE----------

//TRACE VERTEX
trace_vertex.resize(n_trace_vertex); // righe
for(int i = 0; i<n_trace_vertex; i++){
  trace_vertex[i].resize(2); // colonne
}
std::ifstream input4;
input4.open("trace_trace");

if (!input4) {
  std::cout << "Cannot open file.\n";
  return;
}

getline(input4,sGarbage);
for(int i = 0; i<n_trace_vertex; i++){
  input >> garbage;
  input >> trace_vertex[i][0];
  input >> trace_vertex[i][1];
}

//TRACE
trace.resize(n_traces);
for(int i = 0; i < n_traces; i++){
  trace[i].resize(2);
}
getline(input4,sGarbage);
for(int i = 0; i<n_traces; i++){
  input >> garbage;
  input >> trace[i][0];
  input >> trace[i][1];
}
input4.close();

};

void create_additional_structures(){

  // T vettore che per ogni traccia contiene vettore tangente a traccia
  T.resize(n_traces);
  for(int i = 0; i < n_traces; i++){
    T[i].resize(2);
  }

  T_squared.resize(n_traces); // lunghezza al quadrato di ogni T;

  for ( int i = 0; i < n_traces; i++){
    T[i][0]= trace_vertex[trace[i][1]][0] - trace_vertex[trace[i][0]][0];
    T[i][1]= trace_vertex[trace[i][1]][1] - trace_vertex[trace[i][0]][1];
    T_squared[i] = T[i][0]*T[i][0] + T[i][1]*T[i][1];
  }


}

File: 2D/file_reading.m
clear
close all

global n_nodes
global n_edges
global n_triangles
global n_trace_vertex
global n_traces
%la traccia va usata spesso, quindi al posto di passarla in una funzione,
%la mettiamo global.
global trace_vertex
global trace
global T
global node
global edge
global T_squared
global toll_t
global accuracy
global info_trace
global queue
global neigh
global triangle
accuracy = 1e-14;
%---------------------------- NODE --------------------------

fp=fopen('quad.1.node','r');
if (fp==-1)
    disp('Errore apertura file nodi.');
    return;
end
A=fscanf(fp,'%d',4);
n_nodes=A(1);
A=fscanf(fp,'%d %f %f %d',[4 n_nodes]);
node = repmat(struct('x',0,'y',0,'side',0,'toll',-1,'edges',[0,0,0],'triangles',[0,0], ...
              'tot_edges',0,'tot_triangles',0,'s',666),n_nodes,1);
for i=1:n_nodes
    node(i).x = A(2,i);
    node(i).y = A(3,i);
end
%STRUTTURA di 'node': vettore di strutture, creato con repmat
%                     side: da che parte sta rispetto alla traccia in esame
%                           valori possibli: 
%                           0 ----> non sappiamo, 
%              à             1 ----> nodo lato concorde alla normale
%                           -1 ---> nodo lato discorde alla normale
%                           2 ----> nodo sulla retta
%         
%                     s: ascissa curvilinea dell'eventuale intersezione 
%                        con la traccia, se s=666 caso brutto.
%                    
%                     
fclose(fp);

%------------------------- TRIANGLE -------------------------
fp=fopen('quad.1.ele','r');
if (fp==-1)
    disp('Errore apertura file triangoli.');
    return;
end
A=fscanf(fp,'%d',3);
n_triangles=A(1);
triangle = -ones(n_triangles,10);
A=fscanf(fp,'%d',[4 n_triangles]);
triangle(:,[1,2,3])=A([2 3 4],:)';

%STRUTTURA di 'triangle': n_triangles righe x 7 colonne
%                         primi 3 colonne vertici
%                         colonne 4,5,6 status degli edge
%                         colonne 7,8,9 s degli edge
%                         colonna 10 flag:  -1 non sappiamo niente
%                                                -2 lista attesa due nodi
%                                                   in comune
%                                                -3 lista attesa un nodo in
%                                                   comune
%                                                0 tagliato
%                                                >0 posizione su .near_tri
fclose(fp);

%---------------------------- EDGE --------------------------
fp=fopen('quad.1.edge','r');
if (fp==-1)
    disp('Errore apertura file edge.');
    return;
end
A=fscanf(fp,'%d',2);
n_edges=A(1);
A=fscanf(fp,'%d',[4 n_edges]);   
edge=A([2 3],:)';
%STRUTTURA di 'edge': edge(n,1)=vertice 1 del segmento n
%                     edge(n,2)= vertice 2 del segmento n 
fclose(fp);

%--------------------------- NEIGH -------------------------
fp=fopen('quad.1.neigh','r');
if (fp==-1)
    disp('Errore apertura file "neigh".');
    return;
end
A=fscanf(fp,'%d',2);
A=fscanf(fp,'%d',[4 n_triangles]);
neigh=A([2 3 4],:)';
%STRUTTURA di 'neigh': neigh(n,2)=vicino 2 del triangolo n
fclose(fp);

%-------------------------- TRACE -------------------------
fp=fopen('trace.trace','r');
if (fp==-1)
    disp('errore apertura file tracce.');
    return;
end
A=fscanf(fp,'%d',4);
n_trace_vertex=A(1);
n_traces=A(2);
A=fscanf(fp,'%f',[3 n_trace_vertex]);
trace_vertex=A([2 3],:)';
%STRUTTURA di 'trace_vertex': trace_vertex(n,2)=coordinata y del vertice n
fscanf(fp,'%d',2);
A=fscanf(fp,'%d',[3 n_traces]);
trace=A([2 3],:)';
%STRUTTURA di 'trace': trace(n,1)=estremo 1 della traccia n
fclose(fp);


%!!!!!! ------- modifiche di prova -------- !!!!!!
trace_vertex(3,:)=[0.1 0.1];
trace_vertex(4,:)=[0.9 0.9];


%creo matrice T che è la matrice che per ogni traccia contiene 
%il vettore tangente alla traccia.
%il vettore 'T_square' contiene la lunghezza al quadrato di ogni T
T = zeros(n_traces,2);
T_squared = zeros(n_traces,1);

for i=1:n_traces
    T(i,1) = trace_vertex(trace(i,2),1) - trace_vertex(trace(i,1),1);
    T(i,2) = trace_vertex(trace(i,2),2) - trace_vertex(trace(i,1),2);
    T_squared(i) = T(i,1)^2 +T(i,2)^2;
end

%vettore che conterrà la tolleranza per ogni traccia. 
toll_t = zeros(n_traces,1);

%info trace è una struttura che contiene per ogni traccia, 
% -cut_tri = struttura che contiene punti della triangolazione, i due
%            poligoni, i triangoli della triangolazione e id del triangolo
%            tagliato
% - s = vettore di ascisse curvilinee (delle intersezioni con la traccia)
% - near_tri = struttura che contiene i triangoli vicini, in particolare:
%               id : numero triangolo vicino, con i due nodi 

info_trace = repmat(struct('cut_tri',struct('points',zeros(5,2),'poly_1',[],...
                                            'poly_2',[],'tri',zeros(3,3),'id',0),...
                           's',[],...
                           'near_tri',struct('id',0,'nodes',[])),n_traces,1);
 
% queue = coda triangoli da controllare      
% per ogni riga:
%-primo elemento id_tri
%-secondo elemento indice nodo condiviso
%-terzo elemento indice eventuale secondo nodo condiviso 
queue = zeros(0,3);                  


clear A fp